test: Explicitly specify directory where to search tests for

hebasto opened the pull request:
For out-of-source builds, the `test/functional/test_runner.py` is supposed to be run from the build directory which allows it to pick the `test/config.ini` file generated by the build system. Currently, it works accidently for the following reasons:
- on POSIX systems, when running a created by Autoconf symlink to the `test/functional/test_runner.py` in the source directory, it actually has the source directory location in the `sys.path`.
- on Windows (the `build_msvc` directory) VS project puts and copies every build artifact into the source tree (which is wrong and ugly).

This PR makes `test/functional/test_runner.py` work from a build directory in any form (a symbolic link, a hard link, a copy) on _all_ supported platforms, which is highly desirable in the upcoming [CMake-based build system](https://github.com/bitcoin/bitcoin/pull/25797).

For the current master branch, this PR has no behaviour change.

Required for https://github.com/hebasto/bitcoin/pull/15.

---

**Steps to reproduce the issue**

While the issue is mostly specific to Windows and CMake builds, it is still possible to reproduce it on the current master branch.

1. Make an out-of-source build:
```
$ ./autogen.sh
$ mkdir ../build && cd ../build
$ ../bitcoin/configure
$ make
```

2. Note that Autoconf created a symbolic link `test/functional/test_runner.py` in the `../build` directory:
```
$ ls -l test/functional/test_runner.py 
lrwxrwxrwx 1 hebasto hebasto 47 May  5 17:40 test/functional/test_runner.py -> ../../../bitcoin/test/functional/test_runner.py
```
which works flawlessly.

3. However, replacing this symbolic link with a hard link or a copy of `test/functional/test_runner.py` from the source tree will cause the following error:
```
$ cp ../bitcoin/test/functional/test_runner.py test/functional/test_runner.py
$ ls -l test/functional/test_runner.py
$ ./test/functional/test_runner.py 
Temporary test directory at /tmp/test_runner_₿_🏃_20230505_175104
Running Unit Tests for Test Framework Modules
E
======================================================================
ERROR: test_framework (unittest.loader._FailedTest)
----------------------------------------------------------------------
ImportError: Failed to import test module: test_framework
Traceback (most recent call last):
  File "/usr/lib/python3.10/unittest/loader.py", line 154, in loadTestsFromName
    module = __import__(module_name)
ModuleNotFoundError: No module named 'test_framework'


----------------------------------------------------------------------
Ran 1 test in 0.000s

FAILED (errors=1)
Early exiting after failure in TestFramework unit tests
```

hebasto commented:
Friendly ping @stickies-v :)

stickies-v reviewed (commented):
Would it be possible to provide instructions on replicating how to make this fail without`sys.path.append(tests_dir)`? Often times, messing with `system.path` is treating the symptoms when actually the package structure needs to be improved. May not be true in this case, though.

hebasto commented:
@stickies-v 
> Would it be possible to provide instructions on replicating how to make this fail without`sys.path.append(tests_dir)`?

Sure. I've updated the PR description with steps to replicate the failure.

hebasto commented:
Updated a3108880fac89d4472fe3b24d133e9645bed41cf -> fdff2156e1ebc11818f34759bff80e7d37dfff7d ([pr27561.01](https://github.com/hebasto/bitcoin/commits/pr27561.01) -> [pr27561.02](https://github.com/hebasto/bitcoin/commits/pr27561.02), [diff](https://github.com/hebasto/bitcoin/compare/pr27561.01..pr27561.02)):

- addressed @stickies-v's [comment](https://github.com/bitcoin/bitcoin/pull/27561#discussion_r1186191720)

hebasto commented on test/functional/test_runner.py:
Thanks! [Updated](https://github.com/bitcoin/bitcoin/pull/27561#issuecomment-1536535154).

theuni reviewed (commented):
Concept ACK for CMake.

The `path.join` change is easy, but I'll defer to someone more familiar with the test framework on the path change.

maflcko reviewed (approved):
lgtm

hebasto commented:
Dropped non-required refactoring commit.

hebasto commented on test/functional/test_runner.py:
```diff
@@ -536,7 +536,7 @@ def run_tests(*, test_list, src_dir, build_dir, tmpdir, jobs=1, enable_coverage=
         sys.exit(False)
 
-    tests_dir = src_dir + '/test/functional/'
+    tests_dir = os.path.join(src_dir, 'test', 'functional')
```
[Leaved](https://github.com/bitcoin/bitcoin/pull/27561#issuecomment-1554385101) as is.

hebasto commented:
> Didn't check what Cmake does. I presume it copies?

We use [`file(CREATE_LINK ... COPY_ON_ERROR)` ](https://cmake.org/cmake/help/latest/command/file.html#create-link).

stickies-v commented:
> Sure. I've updated the PR description with steps to replicate the failure.

Thanks, this helped a lot with my understanding.

I'm not a huge fan of the build system leaking into the test code, if we can avoid it. Someone reading this without knowledge of the build system wouldn't understand why we're appending to `sys.path`, so at the very least I'd suggest adding some documentation and/or a link to this PR.

An alternative approach could be to add a `bitcoin/test/functional/setup.py` file to make the test framework an installable package. When installing the package in editable mode (`-e`, acts kind of like a symlink), that would prevent developers having to reinstall the package for every change made to the test framework.

<details>
<summary>bitcoin/test/functional/setup.py example:</summary>


```
from setuptools import setup, find_packages

setup(
    name='bitcoin_functional_tests',
    version='0.1',
    packages=find_packages(),
    author='Your Name',
    author_email='your.email@example.com',
    description='Description',
    long_description=open('README.md').read(),
    long_description_content_type='text/markdown',
)
```

</details>

Install the package in editable mode:
```sh
pip install -e ../bitcoin/test/functional/
```

For me, this fixes the problem with the steps you've provided to reproduce the issue. I'm very unfamiliar with the build process though, so this may not be the best solution either. It's definitely more pythonic, and as another upside makes it easier to use the test framework as a library.

stickies-v commented on test/functional/test_runner.py:
```diff
@@ -528,6 +528,10 @@ def run_tests(*, test_list, src_dir, build_dir, tmpdir, jobs=1, enable_coverage=
     print("Running Unit Tests for Test Framework Modules")
+
+    tests_dir = src_dir + '/test/functional/'
+    sys.path.append(tests_dir)
```
Suggested documentation since it's not immediately obvious why this is necessary.
```suggestion
    # This allows `test_runner.py` to work from an out-of-source build directory using a symlink,
    # hard link or a copy on any platform. See https://github.com/bitcoin/bitcoin/pull/27561.
    sys.path.append(tests_dir)
```

hebasto replied:
Thanks! [Updated](https://github.com/bitcoin/bitcoin/pull/27561#issuecomment-1555065798).

stickies-v reviewed (approved):
ACK 342a0c865d8b4b00a088af7b70b1ee0df1864f5c modulo improved documentation (left a suggestion).

> The drawback of such an approach is that it creates additional files in the source tree.

You're right, I can't find a way to install the package without it affecting the source tree in one way or another. So the current approach is probably the best approach.

hebasto committed:
test: Explicitly specify directory where to search tests for

This change allows `test_runner.py` to work from an out-of-source build
directory using a symlink, a hard link or a copy on any platform.

hebasto commented:
Updated 342a0c865d8b4b00a088af7b70b1ee0df1864f5c -> c44f3f231988dc05c4c7a8a96bc2e7b1a54da277 ([pr27561.03](https://github.com/hebasto/bitcoin/commits/pr27561.03) -> [pr27561.04](https://github.com/hebasto/bitcoin/commits/pr27561.04), [diff](https://github.com/hebasto/bitcoin/compare/pr27561.03..pr27561.04)):

- addressed @stickies-v's [comment](https://github.com/bitcoin/bitcoin/pull/27561#discussion_r1199184833)

stickies-v reviewed (approved):
re-ACK https://github.com/bitcoin/bitcoin/commit/c44f3f231988dc05c4c7a8a96bc2e7b1a54da277

maflcko reviewed (commented):
lgtm ACK c44f3f231988dc05c4c7a8a96bc2e7b1a54da277 💸

<details><summary>Show signature</summary>

Signature:

```
untrusted comment: signature from minisign secret key on empty file; verify via: minisign -Vm "${path_to_any_empty_file}" -P RWTRmVTMeKV5noAMqVlsMugDDCyyTSbA3Re5AkUrhvLVln0tSaFWglOw -x "${path_to_this_whole_four_line_signature_blob}"
RUTRmVTMeKV5npGrKx1nqXCw5zeVHdtdYURB/KlyA/LMFgpNCs+SkW9a8N95d+U4AP1RJMi+krxU1A3Yux4bpwZNLvVBKy0wLgM=
trusted comment: lgtm ACK c44f3f231988dc05c4c7a8a96bc2e7b1a54da277 💸
H4uXJb9dCevTXQH5BFAvev56kkzUC6o51P3Mw7DA5MkLu5BfuD8loNRLNiQngaa0PIbX0MUDxQ1sSHObwvcKBQ==
```

</details>
